tests/test-grammar-integration.cpp

```diff
@@ -130,75 +132,161 @@ ws ::= [ \t\n\r]?)""";
 
     // Passing strings
     for (const auto & test_string : test_strings_pass) {
-        auto decoded = decode_utf8(test_string, {});
+        bool matched = match_string(test_string, grammar);
 
-        const auto & code_points = decoded.first;
+        assert(matched);
 
-        int pos = 0;
-        for (auto it = code_points.begin(), end = code_points.end() - 1; it != end; ++it) {
-            ++pos;
-            auto prev_stacks = grammar->stacks;
-            llama_grammar_accept(grammar->rules, prev_stacks, *it, grammar->stacks);
+        // Reset the grammar stacks
+        grammar->stacks = original_stacks;
+    }
 
-            // Expect that each code point will not cause the grammar to fail
-            if (grammar->stacks.empty()) {
-                fprintf(stdout, "Error at position %d\n", pos);
-                fprintf(stderr, "Unexpected character '%s'\n", unicode_cpt_to_utf8(*it).c_str());
-                fprintf(stderr, "Input string is %s:\n", test_string.c_str());
-            }
-            assert(!grammar->stacks.empty());
-        }
+    // Failing strings
+    for (const auto & test_string : test_strings_fail) {
+        bool matched = match_string(test_string, grammar);
 
-        bool completed_grammar = false;
+        assert(!matched);
 
-        for (const auto & stack : grammar->stacks) {
-            if (stack.empty()) {
-                completed_grammar = true;
-                break;
+        // Reset the grammar stacks
+        grammar->stacks = original_stacks;
+    }
+
+    // Clean up allocated memory
+    llama_grammar_free(grammar);
+}
+
+static void test_quantifiers() {
+    // Populate test data with grammar strings and their associated collections of expected passing and failing strings
+    const std::vector<
+        std::tuple<
+            std::string,
+            std::vector<std::string>,
+            std::vector<std::string>>>
```

Comment:
nit: I would suggest using one call per test case instead, so that debugging explodes at the right position / right where the grammar is defined.

```c++
static void test_quantifiers() {
        test_grammar(
            R"""(root ::= "a"*)""",
            // Passing strings
            {
                ...
            },
            // Failing strings
            {
                ...
            }
        });
        ...
```

Reply:
Mmm, that feels better. I'll make that change, thank you!